enhancements/node-healthcheck-operator.md

```diff
@@ -0,0 +1,435 @@
+Check controller except that its creating objects and not
+performing any actions per se. When deploying this controller make
+sure 1. you have a controller implementing the API, e.g [poison pill]
+2. the machine health check controller is not running, or in case of a mixed
+ the node health checks selector avoid targeting nodes with machines.
+
+
+### Risks and Mitigations
+
+#### Remediation Interaction with ClusterAutoScalar
+
+Misconfiguration of nodes provisioning can cause nodes to fail to get operational.
+Remediation is blind to such problems and may interfere with diagnosing the problem.
+The backoff mechanism mentioned in MHC may be suited for Machines, because they
+have the MachineSet name that can be recorded and tracked. Without the machine API
+our knowledge of the nodes could not be handled in such a generic way. There is
+simply no generic way to record a node configuration origin in reliable way without
+putting provider specific annotations.
+Potentially the NHC resource could contain label selectors to track
+repeating remediation and mark them as [incrementally back-off](https://issues.redhat.com/browse/OCPCLOUD-800). e.g
+
+Suggested backoff specification under the NHC resource may look like this:
+```yaml
+kind: NodeHealthCheck
+...
+spec:
+  selector:
+     ...
+  # optional
+  backoff:
+    type: exponential
+    limit: 10m
+```
+
+Caveat of this approach is that we need to take care of conflicting or overlapping backoff definitions.
+Setting global catchall backoff strategy is impossible or just coarse without specifying a selector (how do we know what to backoff from?)
+
+Currently backoff is not implemented.
+
+#### Multiple remediation provider deployed
+
+Multiple providers could act on remediation objects. On non heterogeneous cluster
+this may even make sense. It is not the goal of this controller to coordinate
+those, but providers will need disparate set of nodes to handle based on labels,
+or operator on a node if they can hold a [maintenance lease][node maintenance lease proposal].
+
+
+## Design Details
+
+### Logical Conditions Matching
+
+The resource will be given a list of conditions to track, per a selection of
+nodes, and will evaluate each condition in the list. It's a logical evaluation i.e
+it is enough to evaluate a single condition to true per node to mark as unhealthy.
+
+```
+unhealthyConditions:
+   - type:    "Ready"
+     status:  "Unknown"
+     duration: "300s"
+   - type:    "Ready"
+     status:  "False"
+     duration: "300s"
+```
+In the above example only 1 condition needs to match to mark a node non-healthy.
+
+### Max Unhealthy count
+
+To create remediation object (i.e act for remediation) the overall nodes in the
+selection groups must not exceed `maxUnhealthy` threshold. The value can be
+an int or a percentage, similar to v1/apps `maxUnavailable`:
+```
+spec:
+ maxUnhealthy: 30%
+```
+
+### Default spec values:
```

Comment:
These are the defaults for the NodeHealthCheck spec? Why were these defaults selected?

nit: This section would be easier to read if it was formatted as YAML, or at least wrapped in '```' so the markdown parser does not try to interpret it.

Reply:
They were inherited from the MHC implementation

Reply:
These don't match the MHC defaults. The only one of these fields we actually default is `maxUnhealthy` which defaults to 100% (ie turn off this feature)